Sporadic test failure: TestPurgeWithSomeInvalidDocs

## TestPurgeWithSomeInvalidDocs

Assign to @adamcfraser 

### Repro

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/30/console

### Logs

```
=== RUN   TestPurgeWithSomeInvalidDocs
2017-10-19T13:18:32.407-07:00 Using xattrs: strings.ToLower(useXattrs) == strings.ToLower(TestEnvSyncGatewayTrue).  |true| == |true|
panic: runtime error: slice bounds out of range

goroutine 2213 [running]:
panic(0xbb28a0, 0xc420012100)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/couchbase/sync_gateway/db.parseXattrStreamData(0xc808dc, 0x5, 0xc42090e440, 0x0, 0x0, 0xc3f680, 0xc420748001, 0xc4207405d0, 0xc4206115f8, 0x40857b, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/document.go:194 +0x68b
github.com/couchbase/sync_gateway/db.UnmarshalDocumentSyncDataFromFeed(0xc42090e440, 0x0, 0x0, 0x5, 0xc4207405d0, 0x0, 0xc420611708, 0x58a30d, 0xc420182120, 0xc4206116e7, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/document.go:141 +0x153
github.com/couchbase/sync_gateway/db.(*changeCache).DocChangedSynchronous(0xc4200c21e0, 0x2, 0x0, 0xc42090e440, 0x0, 0x0, 0xc42090e440, 0x0, 0x0, 0x14ef121c2c7a0000, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/change_cache.go:384 +0x1f1
github.com/couchbase/sync_gateway/db.(*changeCache).DocChanged(0xc4200c21e0, 0x2, 0x0, 0xc42090e440, 0x0, 0x0, 0xc42090e440, 0x0, 0x0, 0x14ef121c2c7a0000, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/change_cache.go:340 +0x61
github.com/couchbase/sync_gateway/db.(ChangeIndex).DocChanged-fm(0x2, 0x0, 0xc42090e440, 0x0, 0x0, 0xc42090e440, 0x0, 0x0, 0x14ef121c2c7a0000, 0x50000, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/database.go:225 +0x4c
github.com/couchbase/sync_gateway/db.(*changeListener).ProcessFeedEvent(0xc420496ab0, 0x2, 0x0, 0xc42090e440, 0x0, 0x0, 0xc42090e440, 0x0, 0x0, 0x14ef121c2c7a0000, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/change_listener.go:118 +0x2b3
github.com/couchbase/sync_gateway/db.(*changeListener).ProcessFeedEvent-fm(0x2, 0x0, 0xc42090e440, 0x0, 0x0, 0xc42090e440, 0x0, 0x0, 0x14ef121c2c7a0000, 0x50000, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/db/change_listener.go:92 +0x4a
github.com/couchbase/sync_gateway/base.(*DCPReceiver).DataUpdate(0xc4205f4500, 0x8c, 0xc42090e440, 0x0, 0x0, 0x3, 0xc4205f4800, 0x0, 0x0)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/sync_gateway/base/dcp_feed.go:142 +0x198
github.com/couchbase/go-couchbase/cbdatasource.(*bucketDataSource).worker.func5(0xc4205f9f20, 0xc4201e6500, 0xc420525aa0, 0x26, 0xc420018800, 0x3a, 0xc4206b7740, 0xc4202dace0, 0xc420740540, 0xc4205f9e60, ...)
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/go-couchbase/cbdatasource/cbdatasource.go:1168 +0x8bd
created by github.com/couchbase/go-couchbase/cbdatasource.(*bucketDataSource).worker
	/var/jenkins/workspace/sync-gateway-integration/godeps/src/github.com/couchbase/go-couchbase/cbdatasource/cbdatasource.go:1220 +0x564
exit status 2
FAIL	github.com/couchbase/sync_gateway/rest	91.043s
```

### Notes

From @adamcfraser:

> it was reminding me of https://issues.couchbase.com/browse/MB-25422, but it looks like it's not exactly the same
> similar, maybe?  the datatype is still set as xattr, but there isn't actually any data being sent in the body (because the xattr has been purged)

--- a/base/error.go
+++ b/base/error.go
@@ -30,6 +30,7 @@ const (
 	importCancelledFilter = sgErrorCode(0x05)
 	documentMigrated      = sgErrorCode(0x06)
 	fatalBucketConnection = sgErrorCode(0x07)
+	emptyMetadata         = sgErrorCode(0x08)
 )
 
 type SGError struct {
@@ -45,6 +46,7 @@ var (
 	ErrImportCancelledFilter = &SGError{importCancelledFilter}
 	ErrDocumentMigrated      = &SGError{documentMigrated}
 	ErrFatalBucketConnection = &SGError{fatalBucketConnection}
+	ErrEmptyMetadata         = &SGError{emptyMetadata}
 )
 
 func (e SGError) Error() string {
@@ -65,6 +67,8 @@ func (e SGError) Error() string {
 		return "Timeout performing ViewQuery - could indicate that views are still reindexing"
 	case fatalBucketConnection:
 		return "Fatal error connecting to bucket"
+	case emptyMetadata:
+		return "Empty Sync Gateway metadata"
 	default:
 		return "Unknown error"
 	}

--- a/db/change_cache.go
+++ b/db/change_cache.go
@@ -387,6 +387,9 @@ func (c *changeCache) DocChangedSynchronous(event sgbucket.FeedEvent) {
 		if event.DataType != base.MemcachedDataTypeRaw {
 			base.LogTo("Cache+", "Unable to unmarshal sync metadata for feed document %q.  Will not be included in channel cache.  Error: %v", docID, err)
 		}
+		if err == base.ErrEmptyMetadata {
+			base.Warn("Unexpected empty metadata when processing feed event.  docid: %s opcode: %v datatype:%v", event.Key, event.Opcode, event.DataType)
+		}
 		return
 	}
 

--- a/db/document.go
+++ b/db/document.go
@@ -237,6 +237,10 @@ func UnmarshalDocumentSyncDataFromFeed(data []byte, dataType uint8, needHistory 
 
 func parseXattrStreamData(xattrName string, data []byte) (body []byte, xattr []byte, err error) {
 
+	if len(data) < 4 {
+		return nil, nil, base.ErrEmptyMetadata
+	}
+
 	xattrsLen := binary.BigEndian.Uint32(data[0:4])
 	body = data[xattrsLen+4:]
 	if xattrsLen == 0 {

--- a/db/document_test.go
+++ b/db/document_test.go
@@ -7,6 +7,7 @@ import (
 	"log"
 	"testing"
 
+	"github.com/couchbase/sync_gateway/base"
 	"github.com/couchbaselabs/go.assert"
 )
 
@@ -211,6 +212,12 @@ func TestParseXattr(t *testing.T) {
 	assertNoError(t, err, "Unexpected error parsing dcp body")
 	assert.Equals(t, string(resultBody), string(body))
 	assert.Equals(t, string(resultXattr), "")
+
+	// Attempt to retrieve xattr from empty dcp body
+	emptyBody, emptyXattr, emptyErr := parseXattrStreamData("_sync", []byte{})
+	assert.Equals(t, emptyErr, base.ErrEmptyMetadata)
+	assertTrue(t, emptyBody == nil, "Nil body expected")
+	assertTrue(t, emptyXattr == nil, "Nil xattr expected")
 }
 
 func TestParseDocumentCas(t *testing.T) {

